### PR TITLE
[nnc] Simplify i%100 to i if i is less than 100; fixed #52580

### LINKED ISSUE #52580: [TensorExpr] IRSimplifier can't simplify `i%100` when `i` is known to be less than 100
## Problem Description
IRSimplifier should be able to reason about value ranges of variables to simplify `mod` statements.
For instance, in the following statement
```
{
  for (int i = 0; i < 100; i++) {
    B[i] = A[i % 100];
  }
}
```
`i % 100` can be replaced with `i` since `i` is always less than 100 (it's a loop variable that iterates from 0 to 100).

However, IRSimplifier doesn't do that now.

## Full repro script
```python
import torch

class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C._te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

with kernel_arena_scope():
    dtype = torch._C._te.Dtype.Float
    K = 100
    KK = torch._C._te.ExprHandle.int(K)

    A = torch._C._te.Placeholder('A', dtype, [KK])
    B = torch._C._te.Compute('B', [torch._C._te.DimArg(KK, 'i')],
            lambda k: A.load([k % KK]))

    loopnest = torch._C._te.LoopNest([B])
    print(loopnest.root_stmt())
    loopnest.simplify()
    print('After simplify:', loopnest.root_stmt())
```

### Current behavior
The script outputs
```
Before simplify: {
  for (int i = 0; i < 100; i++) {
    B[i] = A[i % 100];
  }
}
After simplify: {
  for (int i = 0; i < 100; i++) {
    B[i] = A[i % 100];
  }
}
```

### Expected behavior
The script outputs
```
Before simplify: {
  for (int i = 0; i < 100; i++) {
    B[i] = A[i % 100];
  }
}
After simplify: {
  for (int i = 0; i < 100; i++) {
    B[i] = A[i];
  }
}
```

### PR DESCRIPTION
**Pull request by huiguoo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62138
* __->__ #60693
* #60688

Differential Revision: [D29375938](https://our.internmc.facebook.com/intern/diff/D29375938)

### CODE REVIEW

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

File: test/cpp/tensorexpr/test_simplify.cpp [hunk 1]
```diff
@@ -1291,6 +1291,30 @@ TEST(Simplify, SimplifyDivWithLoopContext2) {
...
+TEST(Simplify, SimplifyModWithLoopContext0) {
+  KernelScope kernel_scope;
+  // Stmt to simplify:
+  // for (int i = 0; i < 100; i++) {
+  //  A[i] = i % 100;
+  //}
+  VarHandle i("i", kInt);
+  BufHandle a_buf("A", {100}, kInt);
+  auto forstmt = For::make(
+      i, 0, 100, Store::make(a_buf, {i}, (i % ExprHandle(100))));
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_simplify.cpp:1304]:
```suggestion
      i, 0, 100, Store::make(a_buf, {i}, (i % 100)));
```

File: test/cpp/tensorexpr/test_simplify.cpp [hunk 2]
```diff
@@ -1291,6 +1291,30 @@ TEST(Simplify, SimplifyDivWithLoopContext2) {
...
+
+  const Stmt* simplified = IRSimplifier::simplify(forstmt);
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_simplify.cpp:1306]: Nit: for_stmt

**Review by ZolotukhinM — APPROVED** [review]
LGTM!

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29375938).

**Pull request closed without merging**